### ISSUE TITLE
A yield_now/yield_local alternative that allows the thread to go to sleep

### ISSUE THREAD

**Issue by james7132** [rayon-rs/rayon#1135]
We're investigating using rayon as Bevy's thread pool and scheduler, and part of that requires being able to schedule futures onto the task pool. This has been partially doable with the suggested implementations in the discussion in #679, but we're running into issues where `futures::block_on` equivalents cannot yield to the thread pool in a way that will put the thread to sleep if there is no available work, nor a way to, on-demand, wake that specificthread back up. `futures_lite::block_on` and it's variants park the thread using `parking` or the std equivalent in `std::thread::park`, but rayon does not expose any option for this level of control. Using the aforementioned ways of parking the thread leads to excessive CPU usage and contention, likely due to conflicting with rayon's internal state of how threads should be managed.

**Comment by cuviper** [comment]
Can you sketch what this might look like from the outside?

**Comment by james7132** [comment]
```rust
pub struct Parker;
pub struct Unparker;

impl Parker {
    /// Either does one job, or does an additional round of polling via Sleep, eventually putting the thread to sleep, blocking the
    /// thread until woken up by rayon or the provided Waker.
    pub fn yield_or_sleep(&self) -> Yield {}
}

// The converted Wakes up the paired thread.
impl From<Unparker> for std::task::Waker {
    ...
}

/// Linked to WorkerThread::current
pub fn pair() -> (Parker, Unparker) {}
```
If this is too much to expose to a public interface, it's probably fine to also just have this internally inside `rayon_core` to make a potential `rayon_core::futures::block_on`.

**Comment by cuviper** [comment]
> it's probably fine [...] to make a potential `rayon_core::futures::block_on`

Oh! I had a prototype of that -- here it is dusted off and rebased:
https://github.com/rayon-rs/rayon/compare/main...cuviper:rayon:block_on

**Comment by mtantaoui** [comment]
Hello, I would like to work on this one, do you have a piece od code that I can use to reproduce this one ?

**Comment by U0001F3A2** [comment]
I rebased @cuviper's `block_on` prototype (`cuviper:block_on`) onto current main — the `SpinLatch` `Cow`/`new_static` changes port cleanly, and the `block_on_channel` test still passes.

I dug into the `FIXME: behaves poorly with nesting`. It's rather a hard deadlock, not just poor throughput. Minimal case, linear dependency `A -> outer -> B -> inner`:

- `outer`: `recv(A).await`, then `send(B).await`
- `inner`: `recv(B).await`

`outer` runs and parks waiting on A. While parked, the worker steals `inner` and nests it on the same stack, and `inner` parks waiting on B. Only `outer` sends B, but it can't be re-polled from under the nested `inner` wait, so even once A arrives, nothing progresses. Deadlocks deterministically at 1, 2, and 4 threads; extra workers don't help, since a parked `outer` is a suspended stack frame, not a stealable job. The identical futures complete when `outer` isn't forced to nest (deliver A before `inner` is spawned), so it's the nesting, not the futures.

@cuviper before this becomes a public `rayon_core::block_on`, nesting needs to be prevented or clearly fenced off. Would you rather keep it internal and documented as non-reentrant, or detect re-entrancy (and error / fall back)? Happy to push the rebased branch plus the repro and take it forward whichever way you'd prefer.
